# Домашняя работа №2 (Курс "Advanced Python", 2 семестр)
## ФИО: Реганова Е.В.
11 вариант




**Тема:** SQL  
**Выдана:** 18 марта 2021  
**Мягкий дедлайн:** 25 марта 2021  
**Жесткий дедлайн:** 1 апреля 2021  

**Правила:**
Результат выполнения задания - отчет в формате Jupyter Notebook с кодом и выводами. В ходе выполнения задания требуется реализовать все необходимые алгоритмы, провести эксперименты и ответить на поставленные вопросы. Дополнительные выводы приветствуются. Чем меньше кода и больше комментариев - тем лучше.

Все ячейки должны быть "выполненными", при этом результат должен воспроизвдиться при проверке.

Задание выполняется самостоятельно. **Если вы нашли в Интернете какой-то код, который собираетесь заимствовать, обязательно укажите это в задании.** Если вы советовались с товарищем и/или позаимствовали его решение, обязательно укажите об этом в отчете. Нет ничего плохого в том, что вы пытаетесь разобраться и помогаете друг другу; плохо - когда вы скрываете это и выдаете чужие заслуги за свои. При обнаружении списывания ВСЕМ студентам, имеющим одинаковые списанные решения будет выставлен ОТРИЦАТЕЛЬНЫЙ балл (т.е если задача стоит 4 балла, вы получите не 0, а -4), "оригинал" искаться не будет

Задание, сданное после жесткого дедлайна, не принимается.

## Задание: selectы sql (13 баллов)
Вам предстоит написать несколько Sql запросов к имеющейся базе данных, описывающую King Corporation. Структура базы данных приведена ниже:
![Структура базы данных](scheme.gif "Структура базы данных")

Также,  [здесь](./Таблицы.docx) вы можете посмотреть типы данных столбцов, их смысл, а также sql запросы, которыми можно инициализировать БД такой структуры.

1. Инициализируйте и заполните базу данных, выполнив скрипт [init.sql](./init.sql) **1 балл**  

2. Вставьте в параметр ниже номер вашего варианта и исполните следующие ячейки. Вы получите условия задач вашего варианта. **Каждая задача стоит 4 балла.** Задачи необходимо решить прямо из python при помощи библиотеки sqllite, но если у вас возникли проблемы с использованием sql из python, вы можете просто приложить работающий sql запрос, он будет оценен максимум в 3 балла

In [1]:
MY_VARIANT = 11 #измените этот параметр на номер своего варианта

In [2]:
#Не меняйте эти параметры
TASKS_TO_DO = 3
TOTAL_TASKS = 4

In [2]:
def get_tasks(variant):
    try:
        with open("tasks.txt", encoding="utf8") as tasks:
            for i in range(variant):
                for j in range(TOTAL_TASKS + 1):
                    tasks.readline()
            print(f"Задачи варианта {tasks.readline()}")
            for i in range(TASKS_TO_DO):
                print(i+1, tasks.readline(), end="")
    except:
        print("Что-то пошло не так. Убедитесь, что в текущей рабочей папке"
              " лежит файл tasks.txt и что вы задаете корректный номер варианта")

In [5]:
#выполнение этой функции выведет условия ваших задач. Убедитесь, что вы подгрузили задачи вашего варианта
get_tasks(MY_VARIANT)

Задачи варианта 11

1 Выбрать фамилии и зарплаты тех продавцов (должность - 'SALESPERSON'), которые числятся не в отделах продаж ('SALES')
2 Определить для каждого месяца рекордный объем ежедневных продаж.
3 Выбрать название и маскимальную зарплату+комиссионные по отделу для тех отделов, в которых максимальная зарплата+комиссионные в 1.5 раза больше средней зарплаты по другим отделам фирмы.


### Полезные материалы: 
1. В качестве справки рекомендую читать официальную документацию к Sql от  Microsoft, а также https://professorweb.ru/my/sql-server/2012/level1/ (пункты 2.9, 2.10 будут вам особенно полезны)
2. Также, множество примеров приведены [в файле](./SQL_examples.html). 

Приступайте к написанию sql запросов.

In [17]:
#инициализация базы данных скриптом
import sqlite3
with open('init.sql', 'r') as sql_file:
    sql_script = sql_file.read()
    
con = sqlite3.connect('init.db')
c = con.cursor()
c.executescript(sql_script)

OperationalError: table EMPLOYEE already exists

## Задание1
Выбрать фамилии и зарплаты тех продавцов (должность - 'SALESPERSON'), которые числятся не в отделах продаж ('SALES')

In [27]:

query = ''' SELECT employee.Last_Name, employee.Salary 
                        FROM employee,job, department 
                        WHERE employee.job_id = job.job_id AND employee.department_id = department.department_id
                        AND job.Function = 'SALESPERSON'
                        AND NOT department.name = 'SALES' '''
                       
c.execute(query)
rows = c.fetchall()
for row in rows:
    print(row)

## ТАКИХ НЕТ!!!! 
Ну так-то странно, если продавец не числится в отделе прожаж

## Задание 2
 Определить для каждого месяца рекордный объем ежедневных продаж.

In [51]:
query3 = '''
            WITH dates AS(
            SELECT order_date as date, SUM(total) AS vol
            FROM sales_order
            GROUP BY order_date
            )
            SELECT SUBSTR(date,4,2) as month, MAX(vol) as max
            FROM dates
            GROUP BY month
'''
c.execute(query3)
rows = c.fetchall()
for row in rows:
    print(row)

('01', 5860)
('02', 46370)
('03', 8540)
('04', 3026.5)
('05', 2135.6)
('06', 4900)
('07', 8377.4)
('08', 2409)
('09', 8400)
('10', 3887.5)
('11', 2275.6)
('12', 1620)


## Задание 3

Выбрать название и маскимальную зарплату+комиссионные по отделу для тех отделов, в которых максимальная зарплата+комиссионные в 1.5 раза больше средней зарплаты по другим отделам фирмы.